controllers/nodeupgrade_controller.go

```diff
@@ -2,42 +2,390 @@ package controllers
+	}
+
+	conditions.MarkTrue(nodeUpgrade, anywherev1.UpgraderPodCreated)
+	updateContainerConditions(pod, nodeUpgrade)
+
+	// Always update the readyCondition by summarizing the state of other conditions.
+	conditions.SetSummary(nodeUpgrade,
+		conditions.WithConditions(
+			anywherev1.UpgraderPodCreated,
+			anywherev1.BinariesCopied,
+			anywherev1.ContainerdUpgraded,
+			anywherev1.CNIPluginsUpgraded,
+			anywherev1.KubeadmUpgraded,
+			anywherev1.KubeletUpgraded,
+			anywherev1.PostUpgradeCleanupCompleted,
+		),
+	)
+	return nil
+}
+
+func updateContainerConditions(pod *corev1.Pod, nodeUpgrade *anywherev1.NodeUpgrade) {
+	containersMap := []struct {
+		name      string
+		condition clusterv1.ConditionType
+	}{
+		{
+			name:      upgrader.CopierContainerName,
+			condition: anywherev1.BinariesCopied,
+		},
+		{
+			name:      upgrader.ContainerdUpgraderContainerName,
+			condition: anywherev1.ContainerdUpgraded,
+		},
+		{
+			name:      upgrader.CNIPluginsUpgraderContainerName,
+			condition: anywherev1.CNIPluginsUpgraded,
+		},
+		{
+			name:      upgrader.KubeadmUpgraderContainerName,
+			condition: anywherev1.KubeadmUpgraded,
+		},
+		{
+			name:      upgrader.KubeletUpgradeContainerName,
+			condition: anywherev1.KubeletUpgraded,
+		},
+		{
+			name:      upgrader.PostUpgradeContainerName,
+			condition: anywherev1.PostUpgradeCleanupCompleted,
+		},
+	}
+
+	completed := true
+	for _, container := range containersMap {
+		status, err := getInitContainerStatus(pod, container.name)
+		if err != nil {
+			conditions.MarkFalse(nodeUpgrade, container.condition, "Container status not available yet", clusterv1.ConditionSeverityWarning, "")
+			completed = false
+		} else {
+			if status.State.Waiting != nil {
+				conditions.MarkFalse(nodeUpgrade, container.condition, "Container is waiting to be initialized", clusterv1.ConditionSeverityInfo, "")
+				completed = false
+			} else if status.State.Running != nil {
+				conditions.MarkFalse(nodeUpgrade, container.condition, "Container is still running", clusterv1.ConditionSeverityInfo, "")
+				completed = false
+			} else if status.State.Terminated != nil {
+				if status.State.Terminated.ExitCode != 0 {
+					conditions.MarkFalse(nodeUpgrade, container.condition, fmt.Sprintf("Container exited with a non-zero exit code, reason: %s", status.State.Terminated.Reason), clusterv1.ConditionSeverityError, "")
+					completed = false
+				} else {
+					conditions.MarkTrue(nodeUpgrade, container.condition)
+				}
+			} else {
+				// this should not happen
+				conditions.MarkFalse(nodeUpgrade, container.condition, "Container state is unknown", clusterv1.ConditionSeverityWarning, "")
+				completed = false
+			}
+		}
+	}
+	nodeUpgrade.Status.Completed = completed
+}
+
+func getInitContainerStatus(pod *corev1.Pod, containerName string) (*corev1.ContainerStatus, error) {
+	for _, status := range pod.Status.InitContainerStatuses {
+		if status.Name == containerName {
```

Comment:
We may want a mapping in the future to make it more readable.

Reply:
yeah I can add that in a follow up